emma-language/src/main/scala/org/emmalanguage/compiler/lang/source/Source.scala

```diff
@@ -157,6 +159,101 @@ trait Source extends Common
       //@formatter:on
     }
 
+    trait Algebra[A] {
+
+      // Empty tree
+      def empty: A
+
+      // Atomics
+      def lit(value: Any): A
+      def this_(encl: u.Symbol): A
+      def ref(target: u.TermSymbol): A
+
+      // References (with defaults)
+      def moduleRef(target: u.ModuleSymbol): A = ref(target)
+      def bindingRef(target: u.TermSymbol): A = ref(target)
+      def valRef(target: u.TermSymbol): A = bindingRef(target)
+      def varRef(target: u.TermSymbol): A = bindingRef(target)
+      def parRef(target: u.TermSymbol): A = bindingRef(target)
+
+      // Definitions
```

Comment:
I'm not sure what would be a better alternative. Does DML support recursion? One of the reasons `DefDef`s are missing in `Source` is to prevent the use of recursion. The other is that we use them in `Core` to encode loops. `emma.lib` lets you write non-recursive methods and then inlines them. This brings at least some basic modularity. But I think they need to be in a separate compilation unit :disappointed: